cli/azd/pkg/input/previewer.go

```diff
@@ -0,0 +1,296 @@
+	title string
+	// The title that is displayed on the form of `<prefix>──── <title> ──────────────`.
+	displayTitle string
+	// The line on the bottom. The value is saved on the component state so it is not generated on every re-draw.
+	footerLine string
+	// The number of rows to display for logging.
+	lines int
+	// Use prefix for indentation or any other symbol before each line.
+	prefix string
+	// This list is used as the memory buffer for the logs. The buffer is kept with the size of `lines`
+	output []string
+	// The mutex is used to coordinate updating the header, stopping the component and printing logs.
+	outputMutex sync.Mutex
+}
+
+/****************** Exported method ****************
+- NewProgressLog
+- Start
+- Stop
+- Write
+- Header
+*/
+
+// NewProgressLog returns a new instance of a progressLog.
+func NewProgressLog(lines int, prefix, title, header string) *progressLog {
+	return &progressLog{
+		lines:  lines,
+		prefix: prefix,
+		title:  title,
+		header: header,
+	}
+}
+
+// Start must be call to initialize the component and draw the initial empty frame in the screen.
+// Calling Start() a second time is a no-op. The screen is only updated during the first call.
+// Call Stop() to remove the frame and logs from the screen.
+func (p *progressLog) Start() {
+	if p.output != nil {
+		return
+	}
+	p.output = make([]string, p.lines)
+	// title is created on Start() because it depends on terminal width
+	// if terminal is resized between stop and start, the previewer will
+	// react to it and update the size.
+	p.buildTopBottom()
+
+	// header
+	if p.header != "" {
+		tm.ResetLine("")
+		tm.Println(p.header)
+		// margin after title
+		tm.Println("")
+	}
+
+	// title
+	tm.ResetLine("")
+	tm.Println(p.displayTitle)
+	tm.Println("")
+
+	p.printLogs()
+}
+
+// Stop clears the screen from any previous output and clear the buffer.
+// Calling Stop() before Start() is a no-op.
+func (p *progressLog) Stop() {
+	if p.output == nil {
+		return
+	}
+
+	p.outputMutex.Lock()
+	defer p.outputMutex.Unlock()
+
+	p.clearContentAndFlush()
+
+	// title
+	tm.MoveCursorUp(2)
+	clearLine(p.displayTitle)
+
+	// header
+	if p.header != "" {
+		tm.MoveCursorUp(2)
+		clearLine(p.header)
+	}
+
+	tm.Flush()
+	p.output = nil
+}
+
+// Write implements oi.Writer and updates the internal buffer before flushing it into the screen.
+// Calling Write() before Start() or after Stop() is a no-op
+func (p *progressLog) Write(logBytes []byte) (int, error) {
+	if p.output == nil {
+		return len(logBytes), nil
+	}
+	logsScanner := bufio.NewScanner(strings.NewReader(string(logBytes)))
+	maxWidth := tm.Width()
+	if maxWidth <= 0 {
+		// tm.Width <= 0 means there's no terminal to write and the stdout pipe is mostly connected to a file or a buffer
+		// while azd is been called by another process, like go-test in CI
+		return len(logBytes), nil
+	}
+
+	p.outputMutex.Lock()
+	defer p.outputMutex.Unlock()
+
+	for logsScanner.Scan() {
```

Comment:
The behavior might not be quite right still for `Write("Hello,")` followed by `Write(" World!")`

Reply:
Ideally we'd be able to write some unit tests for this component.  I think it is possible for us to do given that the library we use for the terminal movement and changes allows us to override where it writes to.  We could write to a byte stream and then a small validator that ensures the bytes we wrote to the terminal was correct.

Reply:
I have been using this small test program to play around with how this component works:

```
func main() {
	fmt.Println("---> Starting <---")
	log := input.NewProgressLog(8, "**", "Title", "Header")
	log.Start()
	for i := 0; i < 100; i++ {
		fmt.Fprintf(log, "%d ", i)
		time.Sleep(50 * time.Millisecond)
		if i%10 == 0 {
			fmt.Fprintf(log, "\n")
		}
		if i%20 == 0 {
			log.Header(fmt.Sprintf("Header %d", i))
		}
	}
	time.Sleep(2 * time.Second)
	log.Stop()
	fmt.Println("---> Done <---")
}
```

To @weikanglim's point, it is interesting that the output of this looks like:

```
---> Starting <---
Header 80

**───────── Title ─────────────────────────────────────────────────────────────────────────────────

**92 
**93 
**94 
**95 
**96 
**97 
**98 
**99 

**─────────────────────────────────────────────────────────────────────────────────────────────────
```

I would not expect that a write to the stream without a trailing new line means that we end up writing a newline anyway.

I think the output of this should instead be 8 lines, each with 10 numbers (separated by spaces) (the first two set of 10 will have been scrolled off by the time the program completes).

Reply:
I would like to keep it as such for this first iteration. (write() === writeLine())
There's a few things to define for moving into the model you proposed. Mostly related to what to do on line's overflow. 
On current behavior, we are truncating the line to the terminal's width.

I will create a new issue to design and plan supporting a pure write() model

Reply:
issue: https://github.com/Azure/azure-dev/issues/2452

Reply:
> There's a few things to define for moving into the model you proposed. Mostly related to what to do on line's overflow.
> On current behavior, we are truncating the line to the terminal's width.

I think this behavior is fine, even with a newer model.  Today you are just truncating the line, correct?  It seems like that would be easy to adapt.

I'm sort of concerned that merging this as is yields really weird semantics for previewer, since it ends up inserting arbitrary newlines into the output it's trying to display, which feels wrong.  I'm worried that in practice we're going to end up in a place where output coming into the previewer is not written a line at a time and our output is going to get very wonky.

Reply:
I have update the progress_log component to support sequential writing and added tests cases.

Your sample app should behave now as you are expecting @ellismg .

You will noticed that I removed the `...` for truncating a line that is longer than the screen width.  Instead, the line will continue on the next line.  So, no more truncation.  This make it possible to call `Write(string)` from a loop and use all the space from each line.